public/src/Helpers.ts

```diff
@@ -501,3 +502,98 @@ export class SubmissionsForCourse {
         }
     }
 }
+
+export class SubmissionsForUser {
+    submissions: Map<bigint, Submission[]> = new Map()
+    groupSubmissions: Map<bigint, Submission[]> = new Map()
+    /** ForGroup returns group submissions for the given group */
+    ForGroup(courseID: bigint): Submission[] {
+        return this.groupSubmissions.get(courseID) ?? []
+    }
+
+    ForAssignment(assignment: Assignment): Submission[] {
+        const submissions: Submission[] = []
+        const groupSubs = this.groupSubmissions.get(assignment.CourseID) ?? []
+        const userSubs = this.submissions.get(assignment.CourseID) ?? []
+
+        for (const sub of groupSubs) {
+            if (sub.AssignmentID === assignment.ID) {
+                submissions.push(sub)
+            }
+        }
+
+        for (const sub of userSubs) {
+            if (sub.AssignmentID === assignment.ID) {
+                submissions.push(sub)
+            }
+        }
+        console.log(submissions)
+        return submissions
+    }
+
+    ByID(submissionID: bigint): Submission | undefined {
+        for (const submissions of this.submissions.values()) {
+            const submission = submissions.find(s => s.ID === submissionID)
+            if (submission) {
+                return submission
+            }
+        }
+
+        for (const submissions of this.groupSubmissions.values()) {
+            const submission = submissions.find(s => s.ID === submissionID)
+            if (submission) {
+                return submission
+            }
+        }
+    }
+
+    update(courseID: bigint, submission: Submission) {
```

Comment:
Does this actually update anything? Looks like `subs` should be returned or something. Probably just my lack of typescript knowledge.

Reply:
The `receive` method explicitly updates the array for a given key (courseID), while this method does not.

Since arrays are reference types, using `<Map>.get(courseID)` returns the *reference* to the array for the given course ID. 
Modifying `subs` will therefore modify the same array that is stored in our map.

example:

```js
let a = [1, 2, 3]
let b = a
b[0] = 20
console.log(a) // [20, 2, 3]
```

The method is not being used anywhere however. 
Might be better to remove it and rename `receive` to `update` as `receive` is used to update a submission when the submission stream sends a submission.

Reply:
The reason I thought it wouldn't work was that the `subs` is a `const` which shouldn't be updatable... but maybe you can still update it as a container even though it is a const variable.

But removing it and renaming `receive` is always nice for reducing maintenance overhead.

Reply:
The `const` means I cannot reassign the reference to some other reference, but I can modify the contents of the referenced array.